catalog/app/containers/Bucket/Download/PackageOptions.tsx

```diff
@@ -0,0 +1,184 @@
+  link: {
+    marginBottom: t.spacing(0.5),
+  },
+  copy: {
+    width: 'auto',
+  },
+}))
+
+interface QuiltSyncProps {
+  className?: string
+  uri: PackageUri.PackageUri
+}
+
+function QuiltSync({ className, uri }: QuiltSyncProps) {
+  const classes = useQuiltSyncStyles()
+  const { push } = Notifications.use()
+  const uriString = PackageUri.stringify(uri)
+
+  const handleCopy = React.useCallback(() => {
+    copyToClipboard(uriString)
+    push('URI has been copied to clipboard')
+  }, [uriString, push])
+
+  return (
+    <div className={className}>
+      <M.ButtonGroup variant="outlined" fullWidth className={classes.link}>
+        <M.Button startIcon={<M.Icon>download</M.Icon>} href={uriString}>
+          Open in QuiltSync
+        </M.Button>
+        <M.Button className={classes.copy} onClick={handleCopy}>
+          <M.Icon fontSize="inherit">file_copy_outlined</M.Icon>
+        </M.Button>
+      </M.ButtonGroup>
+      <M.Typography variant="caption">
+        Don't have QuiltSync?{' '}
+        <StyledLink href={urls.quiltSync} target="_blank">
+          Download it here
+        </StyledLink>
+      </M.Typography>
+    </div>
+  )
+}
+
+const useDownloadPanelStyles = M.makeStyles((t) => ({
+  quiltSync: {
+    borderBottom: `1px solid ${t.palette.divider}`,
+    marginBottom: t.spacing(1),
+    paddingBottom: t.spacing(1),
+  },
+}))
+
+interface DownloadPanelProps {
+  fileHandle?: Model.S3.S3ObjectLocation
+  selection?: Selection.ListingSelection
+  uri: Required<Omit<PackageUri.PackageUri, 'tag'>>
+}
+
+function DownloadPanel({ fileHandle, selection, uri }: DownloadPanelProps) {
+  const classes = useDownloadPanelStyles()
+  const { prefs } = BucketPreferences.use()
+  return (
+    <TabPanel>
+      {BucketPreferences.Result.match(
+        {
+          Ok: ({ ui: { actions } }) =>
+            actions.openInDesktop && (
+              <QuiltSync className={classes.quiltSync} uri={uri} />
+            ),
+          _: () => null,
+        },
+        prefs,
+      )}
+      {fileHandle ? (
+        <DownloadFile fileHandle={fileHandle} />
+      ) : (
+        <DownloadDir selection={selection} uri={uri} />
+      )}
+    </TabPanel>
+  )
+}
+
+interface CodePanelProps {
+  hashOrTag: string
+  uri: Required<Omit<PackageUri.PackageUri, 'tag'>>
+}
+
+function CodePanel({ hashOrTag, uri }: CodePanelProps) {
+  return (
+    <TabPanel>
+      <PackageCodeSamples hashOrTag={hashOrTag} {...uri} />
+    </TabPanel>
+  )
+}
+
+interface OptionsProps {
+  fileHandle?: Model.S3.S3ObjectLocation
+  hashOrTag: string
+  hideCode?: boolean
+  selection?: Selection.ListingSelection
+  uri: Required<Omit<PackageUri.PackageUri, 'tag'>>
+}
+
+export default function Options({
+  fileHandle,
+  hashOrTag,
+  hideCode,
+  selection,
+  uri,
+}: OptionsProps) {
+  if (hideCode)
+    return <DownloadPanel fileHandle={fileHandle} selection={selection} uri={uri} />
+
+  return (
+    <Tabs labels={['Download', 'Code']}>
```

Comment:
maybe just make the Tabs component accept `download` and `code`?
that would probably make the code cleaner, bc this `activeTab === 0` thing looks quite confusing.
also, you don't have to make them into functions w/ that approach, just use plain elements.
it would look like this:

```ts
<Tabs
  download={<DownloadPanel fileHandle={fileHandle} selection={selection} uri={uri} />}
  code={<CodePanel hashOrTag={hashOrTag} uri={uri} />}
/>
```

Reply:
Oh, I thought, in `code={<CodePanel hashOrTag={hashOrTag} uri={uri} />}` `CodePanel` will be executed even when not rendered, but it is executed only when rendered

Reply:
I moved labels inside, used `'download' | 'code'` for tabs instead of numbers.

Reply:
> Oh, I thought, in code={<CodePanel hashOrTag={hashOrTag} uri={uri} />} CodePanel will be executed even when not rendered, but it is executed only when rendered

yeah, that's an important thing to know about react:
its model is actually reminiscing monads (🤫) -- you don't execute computations directly, you "describe" them
(you don't call the `CodePanel` function, you just pass it along with the args to the createElement function or smth like that),
and the runtime later executes them in a way it deems necessary.